### ISSUE TITLE
Crash when access Pix colormap on debug

### ISSUE THREAD

**Issue by casensiom** [DanBloomberg/leptonica#743]
What happens if the pix structure has no colormap? is not being check

https://github.com/DanBloomberg/leptonica/blob/e1a26d31310142da81eac20b5cb15fd61057aec9/src/bmpio.c#L567-L571

This code is being crashing when I use tesseract:

```
00007ffb33270ebe() (Unknown Source:0)
pixWriteMemBmp() (Unknown Source:0)
pixWriteMem() (Unknown Source:0)
tesseract::ImageData::SetPixInternal(class tesseract::Image,class std::vector<char,class std::allocator<char> > *) (Unknown Source:0)
tesseract::ImageData::SetPix(class tesseract::Image) (Unknown Source:0)
tesseract::ImageData::ImageData(bool,class tesseract::Image) (Unknown Source:0)
tesseract::Tesseract::GetRectImage(class tesseract::TBOX const &,class tesseract::BLOCK const &,int,class tesseract::TBOX *) (Unknown Source:0)
tesseract::Tesseract::LSTMRecognizeWord(class tesseract::BLOCK const &,class tesseract::ROW *,class tesseract::WERD_RES *,class tesseract::PointerVector<class tesseract::WERD_RES> *) (Unknown Source:0)
tesseract::Tesseract::classify_word_pass1(struct tesseract::WordData const &,class tesseract::WERD_RES * *,class tesseract::PointerVector<class tesseract::WERD_RES> *) (Unknown Source:0)
tesseract::Tesseract::RetryWithLanguage(struct tesseract::WordData const &,void ,bool,class tesseract::WERD_RES * *,class tesseract::PointerVector<class tesseract::WERD_RES> *) (Unknown Source:0)
tesseract::Tesseract::classify_word_and_language(int,class tesseract::PAGE_RES_IT *,struct tesseract::WordData *) (Unknown Source:0)
tesseract::Tesseract::RecogAllWordsPassN(int,class tesseract::ETEXT_DESC *,class tesseract::PAGE_RES_IT *,class std::vector<struct tesseract::WordData,class std::allocator<struct tesseract::WordData> > *) (Unknown Source:0)
tesseract::Tesseract::recog_all_words(class tesseract::PAGE_RES *,class tesseract::ETEXT_DESC *,class tesseract::TBOX const *,char const *,int) (Unknown Source:0)
tesseract::TessBaseAPI::Recognize(class tesseract::ETEXT_DESC *) (Unknown Source:0)

```

**Comment by DanBloomberg** [comment]
BMP writing in leptonica has been tested on many different kinds of images.
For example, the following when run with valgrind has no memory issues, where the
images being tested are all in the prog directory:
```
const char *infile[15] = {"test1.bmp", "speckle2.png", "weasel2.4g.png",
                          "speckle4.png", "dreyfus8.png", "weasel8.240c.png",
                          "test8.jpg", "test16.tif", "marge.jpg",
                          "test32-alpha.png", "test-1bit-alpha.png",
                          "test-cmap-alpha.png", "test-fulltrans-alpha.png",
                          "test-gray-alpha.png", "gray-alpha.tif"};

    pixa1 = pixaCreate(16);
    for (i = 0; i < 15; i++) {
        pix1 = pixRead(infile[i]);
        cmap = pixGetColormap(pix1);
        snprintf(outfile, sizeof(outfile), "/tmp/outfile.%d", i);
        pixWrite(outfile, pix1, IFF_BMP);
        pix2 = pixRead(outfile);
        if (!cmap)
            pix3 = pixRemoveColormap(pix2, REMOVE_CMAP_BASED_ON_SRC);
        else
            pix3 = pixClone(pix2);
        pixaAddPix(pixa1, pix3, L_INSERT);
        pixEqual(pix1, pix3, &equal);
        if (!equal)
            lept_stderr("failure for i = %d\n", i);
        pixDestroy(&pix1);
        pixDestroy(&pix2);
    }
    pix1 = pixaDisplayTiledInColumns(pixa1, 1, 1.0, 30, 2);
    pixWrite("/tmp/results.png", pix1, IFF_PNG);
    pixaDestroy(&pixa1);
    pixDestroy(&pix1);
```

Questions:
* How do you know the failure is due to not checking for a colormap'?
* What is the input image that is failing in tesseract?

It would help to write the pix out in png format.

**Comment by casensiom** [comment]
I'm using a raw image from memory.
I use this method from Tesseract to initialize the image:

```cpp
void TessBaseAPI::SetImage(const unsigned char *imagedata, int width, int height,
                           int bytes_per_pixel, int bytes_per_line) {
  if (InternalSetImage()) {
    thresholder_->SetImage(imagedata, width, height, bytes_per_pixel, bytes_per_line);
    SetInputImage(thresholder_->GetPixRect());
  }
}
```

I found the crash location while debugging. I'm using tesseract 1.83.1

As you can see in the screenshot I have to add an small patch to use the library. (just a check for the presence of colormap)
![leptonica-patch](https://github.com/DanBloomberg/leptonica/assets/7832060/e27b7db0-c26d-48fd-aacd-e32f6c5a8f58)

I will try to write the pix in png format and let you know about it.

**Comment by DanBloomberg** [comment]
Thank you.  You have prevented the crash by adding the check for the colormap, right?

That code in bmpio.c says that bmp files with d <=8 have colormaps.
This pix has d = 32 bpp, with 3 spp.  So it does not have a colormap.
Without the check that you added, it crashes in l. 548 when referencing into NULL.

I don't need the png file and will fix the code.

**Comment by casensiom** [comment]
Right, that’s exactly the situation.
Thanks

**Comment by DanBloomberg** [comment]
Resolved in commit [76c635c](https://github.com/DanBloomberg/leptonica/commit/76c635ceeebcb78ecd22807171c2e2c21c58e)

**Comment by DanBloomberg** [comment]
Thanks again, Carlos !
